server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -2650,6 +2652,28 @@ public long getLocalCheckpoint() {
         return getEngine().getPersistedLocalCheckpoint();
     }
 
+    /**
+     * Fetch the latest checkpoint that has been processed but not necessarily persisted. This should be used only when Segment Replication is enabled.
+     * Also see {@link #getLocalCheckpoint()}.
+     */
+    public long getProcessedLocalCheckpoint() {
+        assert indexSettings.isSegRepEnabled();
+        // Returns checkpoint only if the current engine is an instance of NRTReplicationEngine or InternalEngine
+        return getReplicationEngine().map(NRTReplicationEngine::getProcessedLocalCheckpoint).orElseGet(() -> {
+            final Engine engine = getEngine();
+            assert engine instanceof InternalEngine;
+            return ((InternalEngine) engine).getProcessedLocalCheckpoint();
+        });
+    }
+
+    private Optional<NRTReplicationEngine> getReplicationEngine() {
```

Comment:
This method is not part of actual [PR](https://github.com/opensearch-project/OpenSearch/pull/3470). But to fix the breaking change and write getProcessedLocalCheckpoint() in Indexshard.java in a cleaner way. This method is picked from another Segment Replication [PR](https://github.com/opensearch-project/OpenSearch/pull/3533), which will be backported later. While backporting with [PR](https://github.com/opensearch-project/OpenSearch/pull/3533), we should not include getReplicationEngine() in IndexShard.java